pandas/io/sas/saslib.pyx

```diff
@@ -190,136 +184,202 @@ cdef rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
 
     return np.asarray(outbuff)
 
-cdef decompress(object parser, int row_length, page):
-    page = np.frombuffer(page, dtype=np.uint8)
-    if parser.compression == const.rle_compression:
-        return rle_decompress(row_length, page)
-    elif parser.compression == const.rdc_compression:
-        return rdc_decompress(row_length, page)
-    else:
-        raise ValueError("unknown SAS compression method: %s" %
-                         parser.compression)
+cdef class Parser(object):
 
+    cdef:
+        int column_count
+        long[:] lengths
+        long[:] offsets
+        long[:] column_types
+        uint8_t[:, :] byte_chunk
+        object[:, :] string_chunk
+        char *cached_page
+        int current_row_on_page_index
+        int current_row_in_chunk_index
+        int current_row_in_file_index
+        int row_length
+        int bit_offset
+        int subheader_pointer_length
+        bint is_little_endian
+        np.ndarray[uint8_t, ndim=1] (*decompress)(int result_length, np.ndarray[uint8_t, ndim=1] inbuff)
+        object parser
+
+    def __init__(self, object parser):
+        cdef:
+            int j
+            char[:] column_types
+
+        self.current_row_on_page_index = parser._current_row_on_page_index
+        self.current_row_in_chunk_index = parser._current_row_in_chunk_index
+        self.current_row_in_file_index = parser._current_row_in_file_index
+        self.parser = parser
+        self.column_count = parser.column_count
+        self.lengths = parser._column_data_lengths
+        self.offsets = parser._column_data_offsets
+        self.byte_chunk = parser._byte_chunk
+        self.string_chunk = parser._string_chunk
+        self.row_length = parser.row_length
+        self.cached_page = <char *>parser._cached_page
+        self.bit_offset = self.parser._page_bit_offset
+        self.subheader_pointer_length = self.parser._subheader_pointer_length
+        self.is_little_endian = parser.byte_order == "<"
+        self.column_types = np.empty(self.column_count, dtype=long)
+
+        column_types = parser.column_types
+
+        # map column types
+        for j in range(self.column_count):
+            if column_types[j] == b'd':
```

Comment:
I theory this should be enum types that map to integers but couldn't figure out how to do that efficiently.